### ISSUE TITLE
How to convert image to pdf

### ISSUE THREAD

**Issue by lugeluge** [qpdf/qpdf#383]
in qpdf examples,there is a pdf-create.cc example about how to create pdf and I tried to modify the example to implement the image to pdf function. The jpeg format image is easy to add successfully, but the png and tiff format images have failed. I noticed that there is pngfilter in the library, so someone can provide a example about how to turn png and tiff to pdf

**Comment by jberkenbilt** [comment]
The png filter doesn't have anything to do with png images. png files can use predictors to help with making compression more efficient, exploiting properties of data that are typically found in images, and the PDF spec allows those png filters to be used for arbitrary data. The qpdf library does not contain the code required to directly create PDF files from TIFF or PNG files. JPEG files are special because the stream contents for an image encoded using /DCTDecode happens to be exactly the same as the JPEG file, so making a PDF from a JPEG is trivial. For PNG and TIFF, you would have to understand those image file formats, extract the raw image data, re-encode for PDF, and set up the image dictionary based on stuff you read from headers. An application that did this would most likely link with the PNG or TIFF libraries or some other library, like ImageMagick, that knows how to operate on those image file formats. PNG is relatively straightforward, but TIFF is a lot more complex because there are so many different encodings and layouts supported by TIFF files. Note that, even with jpeg, you have to set up the image dictionary with the correct dimensions and encoding. Also, you could treat jpeg files like any other image file, but doing so would result in degradation of the image quality since jpeg uses lossy compression, and so each decompression/compression cycle decreases image quality. Also, PNG files and even TIFF files *can* contain image data that is properly formatted for PDF, so it's possible that you might be able to take some shortcuts in some cases. It's been a while since I've worked with PNG files, but my recollection is that PNG files usually contain zlib-compressed data. TIFF files sometimes do and sometimes don't. Note that bitonal tiff files (e.g. black and white, not grayscale) often use CCITT group 4 compression, which is supported by PDF, so with sufficient knowledge of the file formats, there are cases in which you could skip the decompress/compress steps for TIFF as well.

Bottom line: with the exception of JPEG where you can take some shortcuts, the procedure would be to read the image data using something aware of the image file format, then re-encode the image for PDF. You won't be able to do it with qpdf alone. Hope that helps.

I'm going to go ahead and close this, but feel free to reopen/comment if you have additional questions.

**Comment by dov** [comment]
On a simiar issue, I have raw RGBA data that I'd like to as an overlay on top of an existing pdf page. Going through JPEG means loosing the alpha channel, since JPEG doesn't support transparency. Is there any way around that?

**Comment by jberkenbilt** [comment]
The way PDF does transparency is not by having an image with an alpha channel but by having an image and associated image mask. This python script, using pikepdf, is a quickie I wrote for my own use a while back. It's not complete on its own, but for someone who is knowledge about PDF, it illustrates how transparency works in a minimalistic way. This script just takes the raw data and sticks in the PDF using deflate (zlib) compression, but there's nothing to stop you from taking an RGBA image, storing the alpha channel as a mask, and converting the rest to jpeg. In fact, you might get that result by running this result through qpdf --optimize-images. There's also a chance that it might convert the mask to jpeg, which is probably not what you want but might still not be that noticeable in some settings.

This is not really end-user ready, but maybe it will be informative to you or someone else who chances upon this issue in the future.


```python
#!/usr/bin/env python3
import argparse
import os
import pikepdf
import PIL.Image as Image
import sys

whoami = os.path.basename(sys.argv[0])


def parse_args(args=sys.argv[1:], prog=whoami):
    parser = argparse.ArgumentParser(
        prog=prog,
        description='Add transparent RGBA PNG a PDF file',
    )
    parser.add_argument('--png',
                        help='path to RGBA PNG file',
                        required=True)
    parser.add_argument('--src',
                        help='original PDF file',
                        required=True)
    parser.add_argument('--dest',
                        help='final PDF file',
                        required=True)
    parser.add_argument('--page',
                        help='page to insert image into, numbered from 1',
                        type=int,
                        required=True)
    parser.add_argument('--x',
                        help='x coordinate of image in points',
                        type=int,
                        required=True)
    parser.add_argument('--y',
                        help='y coordinate of image in points',
                        type=int,
                        required=True)
    parser.add_argument('--width',
                        help='width of image in points',
                        type=int,
                        required=True)
    return parser.parse_args(args)


options = parse_args()

rgba = Image.open(options.png)
(width, height) = rgba.size
rgb = Image.new("RGB", rgba.size, (255, 255, 255))
alpha = rgba.getchannel('A')
rgb.paste(rgba, mask=alpha)

pdf = pikepdf.Pdf.open(options.src)

alpha_stream = pikepdf.Stream(pdf, alpha.tobytes())
alpha_stream['/Type'] = pikepdf.Name('/XObject')
alpha_stream['/Subtype'] = pikepdf.Name('/Image')
alpha_stream['/ColorSpace'] = pikepdf.Name('/DeviceGray')
alpha_stream['/BitsPerComponent'] = 8
alpha_stream['/Interpolate'] = True
alpha_stream['/Width'] = width
alpha_stream['/Height'] = height

rgb_stream = pikepdf.Stream(pdf, rgb.tobytes())
rgb_stream['/Type'] = pikepdf.Name('/XObject')
rgb_stream['/Subtype'] = pikepdf.Name('/Image')
rgb_stream['/ColorSpace'] = pikepdf.Name('/DeviceRGB')
rgb_stream['/BitsPerComponent'] = 8
rgb_stream['/Interpolate'] = True
rgb_stream['/SMask'] = alpha_stream
rgb_stream['/Width'] = width
rgb_stream['/Height'] = height

fxo = pikepdf.Stream(pdf, b'q\n1 0 0 1 0 0 cm\n/Im Do\nQ\n')
fxo['/Type'] = pikepdf.Name('/XObject')
fxo['/Subtype'] = pikepdf.Name('/Form')
fxo['/Resources'] = {
    '/XObject': {
        '/Im': rgb_stream,
    },
}
fxo['/BBox'] = [0, 0, width, height]

scale = options.width / width
im_height = scale * height
prefix = '/FxPNG'

page = pdf.pages[options.page - 1]
if '/Resources' not in page:
    page['/Resources'] = pikepdf.Dictionary(XObject=pikepdf.Dictionary())
elif '/XObject' not in page['/Resources']:
    page['/Resources']['/XObject'] = pikepdf.Dictionary()
xobject = page['/Resources']['/XObject']

n = 1
prefix = None
while prefix is None:
    candidate = f'/FX{n}'
    if candidate not in xobject:
        prefix = candidate
    else:
        n += 1

png_content = f'''Q
q
{options.width} 0 0 {im_height} {options.x} {options.y} cm
{prefix} Do
Q\n'''.encode()
page.page_contents_add(pikepdf.Stream(pdf, 'q\n'.encode()), prepend=True)
page.page_contents_add(pikepdf.Stream(pdf, png_content), prepend=False)
if '/Resources' not in page:
    page['/Resources'] = pikepdf.Dictionary(XObject=pikepdf.Dictionary())
elif '/XObject' not in page['/Resources']:
    page['/Resources']['/XObject'] = pikepdf.Dictionary()
page['/Resources']['/XObject'][prefix] = fxo
pdf.save(options.dest)
```

**Comment by mcp292** [comment]
As I understand it, the answer is to first convert to pdf: `convert image.png image.pdf`.
